src/eigenvalue.cpp

```diff
@@ -0,0 +1,654 @@
+      if (neighbor > mpi::n_procs - 1) break;
+    }
+  }
+
+  // ==========================================================================
+  // RECEIVE BANK SITES FROM NEIGHBORS OR TEMPORARY BANK
+
+  start = simulation::work_index[mpi::rank];
+  index_local = 0;
+
+  // Determine what process has the source sites that will need to be stored at
+  // the beginning of this processor's source bank.
+
+  int neighbor;
+  if (start >= bank_position[mpi::n_procs - 1]) {
+    neighbor = mpi::n_procs - 1;
+  } else {
+    neighbor = upper_bound_index(bank_position, bank_position + mpi::n_procs, start);
+  }
+
+  while (start < simulation::work_index[mpi::rank + 1]) {
+    // Determine how many sites need to be received
+    int64_t n;
+    if (neighbor == mpi::n_procs - 1) {
+      n = simulation::work_index[mpi::rank + 1] - start;
+    } else {
+      n = std::min(bank_position[neighbor + 1], simulation::work_index[mpi::rank + 1]) - start;
+    }
+
+    if (neighbor != mpi::rank) {
+      // If the source sites are not on this processor, initiate an
+      // asynchronous receive for the source sites
+
+      requests.emplace_back();
+      MPI_Irecv(&simulation::source_bank[index_local], static_cast<int>(n), mpi::bank,
+            neighbor, neighbor, mpi::intracomm, &requests.back());
+
+    } else {
+      // If the source sites are on this procesor, we can simply copy them
+      // from the temp_sites bank
+
+      index_temp = start - bank_position[mpi::rank];
+      std::copy(&temp_sites[index_temp], &temp_sites[index_temp + n],
+        &simulation::source_bank[index_local]);
+    }
+
+    // Increment all indices
+    start += n;
+    index_local += n;
+    ++neighbor;
+  }
+
+  // Since we initiated a series of asynchronous ISENDs and IRECVs, now we have
+  // to ensure that the data has actually been communicated before moving on to
+  // the next generation
+
+  int n_request = requests.size();
+  MPI_Waitall(n_request, requests.data(), MPI_STATUSES_IGNORE);
+
+#else
+  std::copy(temp_sites.data(), temp_sites.data() + settings::n_particles,
+    simulation::source_bank.begin());
+#endif
+
+  simulation::time_bank_sendrecv.stop();
+  simulation::time_bank.stop();
+}
+
+void calculate_average_keff()
+{
+  // Determine overall generation and number of active generations
+  int i = overall_generation() - 1;
+  int n;
+  if (simulation::current_batch > settings::n_inactive) {
+    n = settings::gen_per_batch*simulation::n_realizations + simulation::current_gen;
+  } else {
+    n = 0;
+  }
+
+  if (n <= 0) {
+    // For inactive generations, use current generation k as estimate for next
+    // generation
+    simulation::keff = simulation::k_generation[i];
+  } else {
+    // Sample mean of keff
+    simulation::k_sum[0] += simulation::k_generation[i];
+    simulation::k_sum[1] += std::pow(simulation::k_generation[i], 2);
+
+    // Determine mean
+    simulation::keff = simulation::k_sum[0] / n;
+
+    if (n > 1) {
```

Comment:
Optional: It may be worth it here to print a error if `n` is not `> 1`.  Note that this case may be 100% covered by the logic on line `113` (in the "synchronize bank logic") in which case it would be less necessar